atomix/cluster/src/main/java/io/atomix/raft/snapshot/PersistedSnapshotStore.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright © 2020  camunda services GmbH (info@camunda.com)
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package io.atomix.raft.snapshot;
+
+import io.atomix.utils.time.WallClockTimestamp;
+import io.zeebe.util.CloseableSilently;
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * Represents an store, which allows to persist snapshots on a storage, which is implementation
+ * dependent. It is possible to take a transient snapshot, which means you can start taking an
+ * snapshot and can persist it later or abort it. Furthermore it is possible to persist/receive
+ * {@link SnapshotChunk}'s from an already {@link PersistedSnapshot} and persist them in this
+ * current store.
+ *
+ * <p>Only one {@link PersistedSnapshot} at a time is stored in the {@link PersistedSnapshotStore}
+ * and can be received via {@link PersistedSnapshotStore#getLatestSnapshot()}.
+ */
+public interface PersistedSnapshotStore extends CloseableSilently {
+
+  /**
+   * Returns true if a snapshot with the given identifier exists already, false otherwise.
+   *
+   * @param id the snapshot ID to look for
+   * @return true if there is a committed snapshot with this ID, false otherwise
+   * @see SnapshotId
+   */
+  boolean exists(String id);
+
+  /**
+   * Starts a new transient snapshot which can be persisted after the snapshot was taken.
+   *
+   * @param index the index to which the snapshot corresponds to
+   * @param term the term to which the snapshots corresponds to
+   * @param timestamp the time to which the snapshots corresponds to
+   * @return the new transient snapshot
+   */
+  TransientSnapshot newTransientSnapshot(
```

Comment:
I have already two interfaces for receiving and sending for the snapshots, but you want to have two for the store as well to separate it better. Do I get that right?

I like when I can work with just one object and for me it was clean to say "ok I now persist this transient snapshot", which is why the persist is on this interface. The store just keeps the persisted ones and allows to create new transient ones, which can be persisted. But I get that point of having the persist on the store. I have to think about that.

Reply:
I tried it out. I have the feeling that this makes it much more complicated than it needs to be, since I have to maintain also these different types then in raft context. Where it sometimes read only and just wants to check whether there is a snapshot or not, then It needs a storage to receive stuff.

Reply:
You can check this branch if you want https://github.com/zeebe-io/zeebe/tree/zell-multiple-factories